gateway/ln-gateway/src/ng/mod.rs

```diff
@@ -0,0 +1,369 @@
+pub mod pay;
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_stream::stream;
+use bitcoin_hashes::Hash;
+use fedimint_client::derivable_secret::{ChildId, DerivableSecret};
+use fedimint_client::module::gen::ClientModuleGen;
+use fedimint_client::module::ClientModule;
+use fedimint_client::sm::util::MapStateTransitions;
+use fedimint_client::sm::{Context, DynState, ModuleNotifier, OperationId, State};
+use fedimint_client::{
+    sm_enum_variant_translation, Client, DynGlobalClientContext, UpdateStreamOrOutcome,
+};
+use fedimint_core::api::{DynGlobalApi, DynModuleApi};
+use fedimint_core::core::{IntoDynInstance, ModuleInstanceId};
+use fedimint_core::db::{AutocommitError, Database};
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::module::{ExtendsCommonModuleGen, TransactionItemAmount};
+use fedimint_core::{apply, async_trait_maybe_send, Amount, OutPoint, TransactionId};
+use fedimint_ln_client::api::LnFederationApi;
+use fedimint_ln_client::contracts::ContractId;
+use fedimint_ln_common::config::LightningClientConfig;
+use fedimint_ln_common::route_hints::RouteHint;
+use fedimint_ln_common::{
+    ln_operation, LightningCommonGen, LightningGateway, LightningModuleTypes, LightningOutput, KIND,
+};
+use futures::StreamExt;
+use lightning::routing::gossip::RoutingFees;
+use secp256k1::{KeyPair, PublicKey, Secp256k1};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use url::Url;
+
+use self::pay::{GatewayPayCommon, GatewayPayInvoice, GatewayPayStateMachine, GatewayPayStates};
+use crate::gatewaylnrpc::GetNodeInfoResponse;
+use crate::lnrpc_client::ILnRpcClient;
+
+const GW_ANNOUNCEMENT_TTL: Duration = Duration::from_secs(600);
+
+/// The high-level state of a reissue operation started with
+/// [`GatewayClientExt::gateway_pay_bolt11_invoice`].
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum GatewayExtPayStates {
+    Created,
+    Preimage,
+    Success,
```

Comment:
Return `Preimage` here, too. I think this is what the `/pay_invoice` HTTP method will actually be waiting for.